lib/rules/dot-notation.js

```diff
@@ -79,6 +79,8 @@ module.exports = function(context) {
     var options = context.options[0] || {};
     var allowKeywords = options.allowKeywords === void 0 || !!options.allowKeywords;
 
+    var allowPattern = options.allowPattern && new RegExp(options.allowPattern);
```

Comment:
Can you change this to an `if`? In its current form `allowPattern` could be either a boolean or a regular expression, which is a bit weird.